presto-orc/src/main/java/com/facebook/presto/orc/reader/ListSelectiveStreamReader.java

```diff
@@ -0,0 +1,731 @@
+            for (int j = 0; j < elementLengths[i]; j++) {
+                elementPositions[index] = elementOffsets[i] + j;
+                index++;
+            }
+        }
+    }
+
+    private void populateOutputPositionsNoFilter(int elementPositionCount)
+    {
+        if (outputRequired) {
+            elementOutputPositionCount = elementPositionCount;
+            elementOutputPositions = ensureCapacity(elementOutputPositions, elementPositionCount);
+            System.arraycopy(elementPositions, 0, elementOutputPositions, 0, elementPositionCount);
+
+            int offset = 0;
+            for (int i = 0; i < outputPositionCount; i++) {
+                offsets[i] = offset;
+                offset += elementLengths[i];
+            }
+            offsets[outputPositionCount] = offset;
+        }
+    }
+
+    private void populateOutputPositionsWithFilter(int elementPositionCount)
+    {
+        elementOutputPositionCount = 0;
+        elementOutputPositions = ensureCapacity(elementOutputPositions, elementPositionCount);
+
+        indexOutOfBounds = listFilter.getTopLevelIndexOutOfBounds();
+
+        int outputPosition = 0;
+        int elementOffset = 0;
+        boolean[] positionsFailed = listFilter.getTopLevelFailed();
+        for (int i = 0; i < outputPositionCount; i++) {
+            if (!positionsFailed[i]) {
+                // TODO How to raise an error if outputRequired == false
+                indexOutOfBounds[outputPosition] = indexOutOfBounds[i];
+                outputPositions[outputPosition] = outputPositions[i];
+                if (outputRequired) {
+                    if (nullsAllowed && presentStream != null) {
+                        nulls[outputPosition] = nulls[i];
+                    }
+                    offsets[outputPosition] = elementOutputPositionCount;
+                    for (int j = 0; j < elementLengths[i]; j++) {
+                        elementOutputPositions[elementOutputPositionCount] = elementPositions[elementOffset + j];
+                        elementOutputPositionCount++;
+                    }
+                }
+                outputPosition++;
+            }
+            elementOffset += elementLengths[i];
+        }
+        if (outputRequired) {
+            this.offsets[outputPosition] = elementOutputPositionCount;
+        }
+
+        outputPositionCount = outputPosition;
+    }
+
+    @Override
+    public int[] getReadPositions()
+    {
+        return outputPositions;
+    }
+
+    @Override
+    public Block getBlock(int[] positions, int positionCount)
+    {
+        checkArgument(outputPositionCount > 0, "outputPositionCount must be greater than zero");
+        checkState(outputRequired, "This stream reader doesn't produce output");
+        checkState(positionCount <= outputPositionCount, "Not enough values: " + outputPositionCount + ", " + positionCount);
+        checkState(!valuesInUse, "BlockLease hasn't been closed yet");
```

Comment:
This assertion seems to be misplaced

Reply:
@tdcmeehan I can't see why that would be the case. Would you explain some more?